contracts/exchangeIssuance/FlashMintWrapped.sol

```diff
@@ -17,220 +17,1087 @@
+  *
+  * @param _wrapModule                  Address of wrapModule that the integrations in ComponentWrapData are assigned to
+  * @param _wrapData                    ComponentWrapData (underlyingERC20 -> wrapped Set component) for each requiredComponents in the same order
+  *
+  * @return ComponentInvokeWrapData[]   data used for input in issuance methods
+  */
+  function getWrapCallData(
+    address _wrapModule, 
+    ComponentWrapData[] calldata _wrapData
+  ) external view returns(ComponentInvokeWrapData[] memory){
+
+    ComponentInvokeWrapData[] memory invokeWrapData = new ComponentInvokeWrapData[](_wrapData.length);
+
+    for (uint256 i = 0; i < _wrapData.length; ++i) {
+
+      invokeWrapData[i] = ComponentInvokeWrapData({
+        callTarget: address(0),
+        value: 0,
+        callData: ""
+      });
+
+
+      if(_wrapData[i].fromToken == _wrapData[i].toToken) {
+        // no wrapping needed
+        continue;
+      }
+
+      // get the wrap adapter directly from the integration registry
+      IWrapV2Adapter wrapAdapter = IWrapV2Adapter(_getAndValidateAdapter(_wrapData[i].integrationName, _wrapModule));
+
+      // get wrap call info from adapter
+      (invokeWrapData[i].callTarget, invokeWrapData[i].value, invokeWrapData[i].callData) = wrapAdapter
+        .getWrapCallData(
+          _wrapData[i].fromToken,
+          _wrapData[i].toToken,
+          _wrapData[i].amount, // = buyUndlerlyingAmount of ComponentSwapData
+          address(this),
+          _wrapData[i].wrapData
+        );
     }
+  
+    return invokeWrapData;
+  }
+
+  /**
+  * Gets the unwrap callData required for issuance for the passed in ComponentWrapData
+  *
+  * @param _wrapModule                  Address of wrapModule that the integrations in ComponentWrapData are assigned to
+  * @param _unwrapData                  ComponentWrapData (wrapped Set component -> underlyingERC20) for each redeemComponents in the same order
+  *
+  * @return ComponentInvokeWrapData[]   data used for input in redeem methods
+  */
+  function getUnwrapCallData(
+    address _wrapModule, 
+    ComponentWrapData[] calldata _unwrapData
+  ) external view returns(ComponentInvokeWrapData[] memory){
 
-    /**
-     * Runs all the necessary approval functions required before issuing
-     * or redeeming a SetToken. This function need to be called only once before the first time
-     * this smart contract is used on any particular SetToken.
-     *
-     * @param _setToken          Address of the SetToken being initialized
-     * @param _issuanceModule    Address of the issuance module which will be approved to spend component tokens.
-     */
-    function approveSetToken(ISetToken _setToken, address _issuanceModule) external {
+    ComponentInvokeWrapData[] memory invokeUnwrapData = new ComponentInvokeWrapData[](_unwrapData.length);
+
+    for (uint256 i = 0; i < _unwrapData.length; ++i) {
+
+      invokeUnwrapData[i] = ComponentInvokeWrapData({
```

Comment:
Isn't this data the zero data for `ComponentInvokeWrapData` i.e. isn't this initialization unnecessary / wasting gas ?

Reply:
Setting the actual values in line 445 fails if the array element has not been initialized before, but it should work with just 
`invokeUnwrapData.length++`. Not sure if that saves gas, but it could. I can try that

Or did you have something else in mind?